lib/src/pubspec.dart

```diff
@@ -117,7 +117,28 @@ class Pubspec {
   SdkConstraintStatus get sdkConstraintStatus =>
       SdkConstraintStatus.fromSdkVersion(_inner.environment?['sdk']);
 
-  VersionConstraint? get dartSdkConstraint => _inner.environment?['sdk'];
+  VersionConstraint? get dartSdkConstraint {
+    final constraint = _inner.environment?['sdk'];
+    // If a package is null safe it should also be compatible with dart 3.
+    // Therefore we rewrite a null-safety enabled constraint with the upper
+    // bound <3.0.0 to be have upper bound <4.0.0
+    if (constraint is VersionRange &&
+        constraint.min != null &&
+        isNullSafety(constraint.min!) &&
+        // <3.0.0 is parsed into a max of 3.0.0-0, so that is what we look for
+        // here.
+        constraint.max == Version(3, 0, 0).firstPreRelease &&
```

Comment:
Nit: extract `Version(3, 0, 0).firstPreRelease` into a separate (final) variable?
Also for the `Version(4, 0, 0)` below..

Reply:
I extracted `Version(3, 0, 0)`, that makes the comments easier to follow.